h2o-core/src/main/java/water/H2ONode.java

```diff
@@ -164,66 +189,59 @@ public boolean isHealthy(long now) {
 
   static H2ONode getClientByIPPort(String ipPort){
     for( Map.Entry<H2Okey, H2ONode> entry : INTERN.entrySet()){
-      if (entry.getValue()._client && entry.getValue().getIpPortString().equals(ipPort)) {
+      if (entry.getValue().isClient() && !entry.getValue().isRemovedFromCloud() && entry.getValue().getIpPortString().equals(ipPort)) {
         return entry.getValue();
       }
     }
     return null;
   }
 
-  private synchronized void refreshClient(short timestamp) {
-    assert timestamp != 0 && H2O.decodeIsClient(timestamp);
-
-    UDP_TCP_SendThread oldSendThread = _sendThread;
-
-    if (_timestamp != 0) {
-      Log.info("Client reconnected with a new timestamp=" + _timestamp + ", old client: " + toDebugString());
+  private void refreshClient(short timestamp) {
+    boolean reconnected = _timestamp != 0 && timestamp != 0 && _timestamp != timestamp;
+    if (reconnected) {
+      Log.info("Client reconnected with a new timestamp=" + timestamp + ", old client: " + toDebugString());
+      if (_sendThread != null) {
```

Comment:
the `if` blocks seems unnecessary for me here as initialization of `sentThread` is already handled in `sentMessage method`

Reply:
We assume a lost client will eventually re-connect and we will want to deliver all the messages - see isActive method in the SendThread. However, when we detect a client was re-spawned (and lost its previous state). We don't need to keep sending the old messages. That is why I kill the old thread right away by injecting a new fresh instance. The old thread will detect it is not active and will give the new one chance to start delivering messages.

Reply:
Oki, thanks, makes sense. So basically sentMessage is just the place where sentThread is lazily initialized if null.

A small comment in the code might help. This small if has a lot of reasoning behind it :)

Reply:
thanks, I will add this to the code